### PR TITLE
Fixing arm compilation.

### PR DESCRIPTION
**Pull request by Layer3**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -25,6 +34,14 @@ download_dep(tinyply
         GIT_TAG point_cloud_utils
 )
 
+if(NOT NATIVE_SSE)
+  # igl has sse hardcoded, so use simde to redirect to the platforms native simd instructions.
+  download_dep(simde
```
Comment (fwilliams) [inline CMakeLists.txt:39]: Could you add version tag or commit ID here so we don't get build errors in the future if there's a bug that gets introduced in the main branch of simde? 

Looks like the [latest stable release is here](https://github.com/simd-everywhere/simde/releases/tag/v0.7.2)

File: external/geogram/cmake/platforms/Linux64-clang/config.cmake [hunk 2]
```diff
@@ -1,5 +1,9 @@
...
 include(${GEOGRAM_SOURCE_DIR}/cmake/platforms/Linux-clang.cmake)
-add_flags(CMAKE_CXX_FLAGS -m64)
-add_flags(CMAKE_C_FLAGS -m64)
 
+IF(${CMAKE_SYSTEM_PROCESSOR} MATCHES "arm")
+    add_flags(CMAKE_CXX_FLAGS -m64)
+    add_flags(CMAKE_C_FLAGS -m64)
+ELSE(${CMAKE_SYSTEM_PROCESSOR} MATCHES "arm")
+    message("NOT ON ARM CPU! Omitting arm specific flag.")
+ENDIF(${CMAKE_SYSTEM_PROCESSOR} MATCHES "arm")
```
Comment (fwilliams) [inline external/geogram/cmake/platforms/Linux64-clang/config.cmake:9]: Is this omitting -m64 when not on ARM?

This worked before on x86 without omitting arm specific flags right. It seems that now when you are compiling on ARM, the behavior is the same as before, but on x86 there is no -m64/-m32. Do we need to change these?

File: external/geogram/cmake/platforms/Linux64-mic-icc/config.cmake [hunk 3]
```diff
@@ -3,5 +3,3 @@ include(${GEOGRAM_SOURCE_DIR}/cmake/platforms/Linux-mic-icc.cmake)
 add_flags(CMAKE_CXX_FLAGS -mmic)
 add_flags(CMAKE_C_FLAGS -mmic)
 set(VORPALINE_WITH_GRAPHICS FALSE)
```
Comment (fwilliams) [inline external/geogram/cmake/platforms/Linux64-mic-icc/config.cmake:5]: Nit: No need to rm newlines here

**Review by fwilliams — COMMENTED** [review]
This is great, thank you so much for doing this!

I pushed a CI chance to your branch to build wheels and run test on ARM architectures on Linux and MacOS (if MacOS fails, we can just do Linux for now and that's fine).

Taking a closer look at simde, it seems there is no performance overhead in just using it directly. I wonder if we could skip the NATIVE_SSE flag and just always use simde. I guess we'll need to check where else uses SSE instructions.

File: external/geogram/cmake/platforms/Linux64-mic-icc/config.cmake (hunk 3 above)
Comment (Layer3) [reply external/geogram/cmake/platforms/Linux64-mic-icc/config.cmake:5]: Reverting this.

File: external/geogram/cmake/platforms/Linux64-clang/config.cmake (hunk 2 above)
Comment (Layer3) [reply external/geogram/cmake/platforms/Linux64-clang/config.cmake:9]: Yes, you are right and this is giving me a little bit of a headache.
I have tested this specifically only on the Khadas Vim3 pro board, which has arm Cortextes, but is running a custom Linux setup that is being developed by the manufacturer.
So I think on some levels/flags the Cmake will tell you that it is x86, but on others it will show arm. I think here it just does not match the `CMAKE_SYSTEM_PROCESSOR` with arm, because on that Linux system the processor doesn't have a specifier string, just a manufacturer- and type/generation-id.

TL;DR: this should actually be the other way around, because m32/64 are i386 specifics.
I think the most sensible solution for now is to switch around these if/else branches and add `NOT NATIVE_SSE` to them. 
What do you think @fwilliams

File: CMakeLists.txt (hunk 1 above)
Comment (Layer3) [reply CMakeLists.txt:39]: AyeAye.
I'd like to add the latest commit for now, as that is the one that we are also using, so it is getting the most coverage.
I can keep an eye on this and switch to a stable release if we run into any issues while testing.

File: external/geogram/cmake/platforms/Linux64-clang/config.cmake (hunk 2 above)
Comment (Layer3) [reply external/geogram/cmake/platforms/Linux64-clang/config.cmake]: With latest this compiles both on the arm test-board (with `NATIVE_SSE OFF`) and on my desktop Linux on x86.

**Comment by Layer3** [comment]
Sorry for taking so long to follow up on all of this.

> I pushed a CI chance to your branch to build wheels and run test on ARM architectures on Linux and MacOS (if MacOS fails, we can just do Linux for now and that's fine).

This shouldn't change anything on Mac anymore, but not entirely sure. I'd leave this completely up to you. I don't own a mac platform, though I might get one if the number of projects requiring mac support continues following this trend for me.
 
> Taking a closer look at simde, it seems there is no performance overhead in just using it directly. I wonder if we could skip the NATIVE_SSE flag and just always use simde. I guess we'll need to check where else uses SSE instructions.

Definitely! The big problem with this imo is that it definitely requires more setup and testing, but the gain is there.
As far as I've seen this requires some more involved changes to the simd abstraction in geogram. I'm not opposed to haivng a look at this, just saying that it is definitely a bit of work.

One more point about the `NATIVE_SSE` flag. Currently this feels more like a `BUILD_KHADAS` flag that has the added benefit of having a high chance to work on other platforms that might have problems to compile with sse.
The actual issue we are having is that Khadas-Linux isn't playing perfectly nice with CMake I think.
Also in Geogram this might be resolved differently, by adding an additional platform for arm Linux and make sure that arm boards switch to that configuration instead of defaulting to linux-gcc-dynamic.

Sorry for the novel, hope I'm not stating the obvious too often. : )

**Comment by fwilliams** [comment]
Okay I think I fixed the CI and merged these changes. I'll try and get #55 working on top of these changes!

**Comment by Layer3** [comment]
Cool, thank you!
